worker/provisioner/provisioner_task.go

```diff
@@ -689,83 +706,313 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
 	return nil
 }
 
-func (task *provisionerTask) startMachines(machines []*apiprovisioner.Machine) error {
-	for _, m := range machines {
-		// Make sure we shouldn't be stopping before we start the next machine
-		select {
-		case <-task.catacomb.Dying():
-			return task.catacomb.ErrDying()
-		default:
-		}
+type availabilityZoneMachine struct {
+	zoneName   string
+	machineIds set.Strings
+	failures   int
```

Comment:
To confirm, you're talking about failures in StartInstance() which happen before the loop on availabilityZones?

Currently the provisioner retries on any error from StartInstance(), provided that the catacomb isn't dying and all of the attempts haven't been used up.  The Providers are retrying with a different AZ on individual criteria.

I was thinking of creating an error for StartInstance() specifically related to AZ failures.  Will do that so we can really know if an AZ is failing.

Tracking failures at the provisioner level aids in suggesting a different AZ to the provider if the first one fails. Also, there are cases where AZs are marked available by the cloud, but always fail to start an instance, I wanted a way to avoid trying them.  But the algorithm around this definitely needs work.